src/paths/steiner.cpp

```diff
@@ -0,0 +1,650 @@
+			std::advance(it, i);
+			std::set<igraph_integer_t> D = *it;
+			igraph_integer_t indexOfSubsetD;
+			indexOfSubsetD = fetchIndexofMapofSets(D);
+
+			for (igraph_integer_t j = 0; j < no_of_vertices; j++)
+			{
+				MATRIX(dp_cache, indexOfSubsetD, j) = IGRAPH_INFINITY;
+			}
+
+			for (igraph_integer_t j = 0; j < no_of_vertices; j++)
+			{
+				igraph_real_t distance1 = IGRAPH_INFINITY;
+				std::set<igraph_integer_t>::iterator subset_D_iterator;
+
+				for (subset_D_iterator = D.begin(); subset_D_iterator != D.end(); subset_D_iterator++)
+				{
+					igraph_integer_t E = *subset_D_iterator;
+					if (E != j)
+					{
+						igraph_integer_t distanceEJ = MATRIX(distance, E, j);
+
+						std::set<igraph_integer_t> DMinusE = D;
+
+						/*
+						 *  A set with Singleton value E removed from subset D
+						 */
+						for (std::set<igraph_integer_t>::iterator iter = DMinusE.begin(); iter != DMinusE.end();)
+						{
+							if (*iter == E)
+							{
+								iter = DMinusE.erase(iter);
+								break;
+							}
+							++iter;
+						}
+
+						igraph_integer_t indexOfSubsetDMinusE;
+						if (DMinusE.size() == 1)
+						{
+							std::set<igraph_integer_t>::iterator node = DMinusE.begin();
+							indexOfSubsetDMinusE = *node;
+						}
+						else
+						{
+							indexOfSubsetDMinusE = fetchIndexofMapofSets(DMinusE);
+						}
+
+						if ((distanceEJ + MATRIX(dp_cache, indexOfSubsetDMinusE, j)) < distance1)
+						{
+							distance1 = distanceEJ + (MATRIX(dp_cache, indexOfSubsetDMinusE, j));
+						}
+					}
+				}
+
+				for (igraph_integer_t k = 0; k < no_of_vertices; k++)
+				{
+					MATRIX(dp_cache, indexOfSubsetD, k) = std::min(MATRIX(dp_cache, indexOfSubsetD, k), MATRIX(distance, k, j) + distance1);
+				}
+			}
+		}
+	}
+
+	igraph_real_t distance2 = IGRAPH_INFINITY;
+
+	for (igraph_integer_t j = 0; j < no_of_vertices; j++)
+	{
+		igraph_real_t distance1 = IGRAPH_INFINITY;
+		for (igraph_integer_t subset_C_iterator = 0; subset_C_iterator < igraph_vector_int_size(steiner_terminals); subset_C_iterator++)
+		{
+			igraph_integer_t F = VECTOR(steiner_terminals_copy)[subset_C_iterator];
+			igraph_integer_t distanceFJ = MATRIX(distance, F, j);
+
+			std::set<igraph_integer_t> CMinusF;
+
+			for (igraph_integer_t k = 0; k < igraph_vector_int_size(steiner_terminals); k++)
+			{
+
+				if (VECTOR(steiner_terminals_copy)[k] != F)
+				{
+					CMinusF.insert(VECTOR(steiner_terminals_copy)[k]);
+				}
+			}
+
+			igraph_integer_t indexOfSubsetCMinusF = fetchIndexofMapofSets(CMinusF);
+
+			if (distanceFJ != 0 && (distanceFJ + (MATRIX(dp_cache, indexOfSubsetCMinusF, j)) < distance1))
+			{
+				distance1 = distanceFJ + (MATRIX(dp_cache, indexOfSubsetCMinusF, j));
+			}
+		}
+
+		if (q != j && MATRIX(distance, q, j) + distance1 < distance2)
+		{
+			distance2 = MATRIX(distance, q, j) + distance1;
+		}
+	}
+	*res = distance2;
+
+	std::set<igraph_integer_t> newSet;
+	for (auto i = 0; i < igraph_vector_int_size(&steiner_terminals_copy); i++)
```

Comment:
`auto` takes on the type of the initialized here, which is `0`, thus it has type `int`. All integers that are used to index into igraph vectors should be `igraph_integer_t`.